concurrency/executor.go

```diff
@@ -0,0 +1,212 @@
+// Executor will attempt to distribute the tasks evenly among the executors.
+func (e *ConcurrentExecutor[ResultType, ResultChannelType, TaskType]) Execute(concurrency int, payload []TaskType, processorFn TaskProcessorFn[ResultChannelType, TaskType]) ([]ResultType, error) {
+	if len(payload) == 0 {
+		return []ResultType{}, nil
+	}
+
+	if concurrency <= 0 {
+		e.logger.Warn().Msg("Concurrency is less than 1, setting it to 1")
+		concurrency = 1
+	}
+
+	var wgProcesses sync.WaitGroup
+	wgProcesses.Add(len(payload))
+
+	canSafelyContinueCh := make(chan struct{}) // waits until listening goroutine finishes, so we can safely return from the function
+	doneProcessingCh := make(chan struct{})    // signals that both result and error channels are closed
+	errorCh := make(chan error, len(payload))
+	resultCh := make(chan ResultChannelType, len(payload))
+
+	// mutex to protect shared state
+	mutex := sync.Mutex{}
+
+	// atomic counter to keep track of processed tasks
+	var atomicCounter atomic.Int32
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	// listen in the background until all tasks are processed (no fail-fast)
+	go func() {
+		defer func() {
+			e.logger.Trace().Msg("Finished listening to task processing results")
+			close(canSafelyContinueCh)
+		}()
+		for {
+			select {
+			case err, ok := <-errorCh:
+				if !ok {
+					e.logger.Trace().Msg("Error channel closed")
+					return
+				}
+				if err != nil {
+					mutex.Lock()
+					e.errors = append(e.errors, err)
+					e.logger.Err(err).Msg("Error processing a task")
+					mutex.Unlock()
+					wgProcesses.Done()
+
+					// cancel the context if failFast is enabled and it hasn't been cancelled yet
+					if e.failFast && ctx.Err() == nil {
+						cancel()
+					}
+				}
+			case result, ok := <-resultCh:
+				if !ok {
+					e.logger.Trace().Msg("Result channel closed")
+					return
+				}
+
+				counter := atomicCounter.Add(1)
+				mutex.Lock()
+				e.results = append(e.results, result.GetResult())
+				e.logger.Trace().Str("Done/Total", fmt.Sprintf("%d/%d", counter, len(payload))).Msg("Finished aggregating task result")
+				mutex.Unlock()
+				wgProcesses.Done()
+			case <-doneProcessingCh:
+				e.logger.Trace().Msg("Signaling that processing is done")
+				return
+			}
+		}
+	}()
+
+	dividedPayload := slice.DivideSlice(payload, concurrency)
+
+	for executorNum := 0; executorNum < concurrency; executorNum++ {
+		go func(key int) {
+			payloads := dividedPayload[key]
+
+			if len(payloads) == 0 {
+				return
+			}
+
+			e.logger.Debug().
+				Int("Executor Index", key).
+				Int("Tasks to process", len(payloads)).
+				Msg("Started processing tasks")
+
+			for i := 0; i < len(payloads); i++ {
+
+				// if context is cancelled and failFast is enabled  mark all remaining tasks as finished
+				if e.failFast && ctx.Err() != nil {
+					e.logger.Trace().
+						Int("Executor Index", key).
+						Str("Cancelled/Total", fmt.Sprintf("%d/%d", (i+1), len(payloads))).
```

Comment:
does i+1 not denote completed task here?

Reply:
that depends on the context. In this loop it might be both:
* `0` - processed all right
CANCEL
* `1` - cancelled task
* `2` - cancelled task

That log here is just for visibility or debugging, if we end up in an unlikely situation, when processing never finishes, it will be easier to figure out which executor failed to mark taks as done.